### PR TITLE
refac: Split sum_within_month test into smaller test

### PR DESCRIPTION
**Pull request by johevemi**
## References

* #0000

### CODE REVIEW

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py [hunk 1]
```diff
@@ -339,34 +340,88 @@ def test__calculate_tariff_price_per_ga_co_es__rounds_total_amount_correctly(
...
+        calculate_tariff_price_per_ga_co_es(tariffs), DEFAULT_TIME_ZONE
+    )
+
+    # Assert
+    assert actual.collect()[0][Colname.total_amount] == Decimal("6.030015")
+    assert actual.collect()[1][Colname.total_amount] == Decimal("2.010005")
+    assert actual.count() == 2
+
+
+def test__sum_within_month__group_by_does_not_care_about_metering_point_type(
```
Comment (BjarkeMeier) [inline source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:366]:
```suggestion
def test__sum_within_month__sums_across_metering_point_types(
```

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py [hunk 2]
```diff
@@ -339,34 +340,88 @@ def test__calculate_tariff_price_per_ga_co_es__rounds_total_amount_correctly(
     assert actual_amount == expected_total_amount
 
 
-def test__sum_within_month__on_tariff(
+def test__sum_within_month__groups_to_month(
```
Comment (BjarkeMeier) [inline source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:343]:
```suggestion
def test__sum_within_month__creates_sum_per_month(
```

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py [hunk 3]
```diff
@@ -339,34 +340,88 @@ def test__calculate_tariff_price_per_ga_co_es__rounds_total_amount_correctly(
...
         _create_tariff_hour_row(charge_time=datetime(2020, 1, 1, 1)),
-        _create_tariff_hour_row(
-            charge_time=datetime(2020, 1, 1, 0), quality=ChargeQuality.ESTIMATED
-        ),
-        _create_tariff_hour_row(charge_time=datetime(2019, 12, 31, 23)),
-        _create_tariff_hour_row(
-            charge_time=datetime(2020, 1, 1, 2),
-            metering_point_type=MeteringPointType.PRODUCTION,
-        ),
+        _create_tariff_hour_row(charge_time=datetime(2020, 1, 1, 0)),
```
Comment (BjarkeMeier) [inline source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:349]: Isn't two rows enough for this test case?

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py [hunk 4]
```diff
@@ -339,34 +340,88 @@ def test__calculate_tariff_price_per_ga_co_es__rounds_total_amount_correctly(
...
+        calculate_tariff_price_per_ga_co_es(tariffs), DEFAULT_TIME_ZONE
+    )
+
+    # Assert
+    assert actual.collect()[0][Colname.qualities] == ["calculated", "estimated"]
+    assert actual.collect()[0][Colname.total_amount] == Decimal("4.020010")
+    assert actual.count() == 1
+
+
+def test__sum_within_month__use_danish_time_to_group(
```
Comment (BjarkeMeier) [inline source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:407]:
```suggestion
def test__sum_within_month__groups_by_local_time_months(
```

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py
Comment (BjarkeMeier) [inline source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:1]: Should DST be tested as well?

Comment (BjarkeMeier) [inline source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:1]: I think that probably all columns from the `.agg(...)` function of the sut deserve test cases.

Comment (johevemi) [reply source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:1]: yes, i have create two tests

Comment (johevemi) [reply source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:1]: total_quantity, charge_price, and unit are irrelevant to the message that needs to be sent out. That said I can make tests for them if that is something we want anyways

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py (hunk 3 above)
Comment (johevemi) [reply source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:349]: I have taken it down to 3 rows, I like seeing the total_amount change

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py
Comment (johevemi) [reply source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:1]: added test for total_quantity and charge_price

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py (hunk 3 above)
Comment (BjarkeMeier) [reply source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:349]: Isn’t that another test case?

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py [hunk 5]
```diff
@@ -339,34 +340,179 @@ def test__calculate_tariff_price_per_ga_co_es__rounds_total_amount_correctly(
     assert actual_amount == expected_total_amount
 
 
-def test__sum_within_month__on_tariff(
+def test__sum_within_month__creates_sum_per_month(
```
Comment (BjarkeMeier) [inline source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:343]: A number of things are summed but just one is tested here
```suggestion
def test__sum_within_month__sums_amount_per_month(
```

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py [hunk 6]
```diff
@@ -339,34 +340,179 @@ def test__calculate_tariff_price_per_ga_co_es__rounds_total_amount_correctly(
...
+
+    # Assert
+    assert actual.collect()[0][Colname.total_amount] == Decimal("4.020010")
+    assert actual.collect()[1][Colname.total_amount] == Decimal("2.010005")
+    assert actual.collect()[0][Colname.charge_time] == datetime(2020, 10, 31, 21)
+    assert actual.collect()[1][Colname.charge_time] == datetime(2020, 10, 31, 23)
+    assert actual.count() == 2
+
+
+def test__sum_within_month__creates_sum_total_quantity(
```
Comment (BjarkeMeier) [inline source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:477]: Perhaps align naming of these "create sum" test cases and put them next to each other?
```suggestion
def test__sum_within_month__sums_quantity_per_month(
```

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py [hunk 7]
```diff
@@ -339,34 +340,179 @@ def test__calculate_tariff_price_per_ga_co_es__rounds_total_amount_correctly(
...
+    actual = sum_within_month(
+        calculate_tariff_price_per_ga_co_es(tariffs), DEFAULT_TIME_ZONE
+    )
+
+    # Assert
+    assert actual.collect()[0][Colname.total_quantity] == Decimal("2.222")
+    assert actual.count() == 1
+
+
+def test__sum_within_month__creates_sum_charge_price(
```
Comment (BjarkeMeier) [inline source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:497]:
```suggestion
def test__sum_within_month__sums_charge_price_per_month(
```

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py [hunk 8]
```diff
@@ -339,34 +340,179 @@ def test__calculate_tariff_price_per_ga_co_es__rounds_total_amount_correctly(
...
-    assert actual.collect()[1][Colname.total_amount] == Decimal("8.040020")
-    assert actual.collect()[2][Colname.total_amount] == Decimal("2.010005")
-    assert actual.collect()[1][Colname.qualities] == ["calculated", "estimated"]
+    assert actual.collect()[1][Colname.total_amount] == Decimal("4.020010")
     assert actual.collect()[1][Colname.charge_time] == datetime(2019, 12, 31, 23)
-    assert actual.count() == 3
+    assert actual.count() == 2
+
+
+def test__sum_within_month__daylight_savings_time_summertime(
```
Comment (BjarkeMeier) [inline source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:429]: I wasn't sure when I suggested to perhaps test for DST. I'm not sure exactly what these tests should test regarding DST. And I'm also - as a second thought - leaning towards not writing DST tests as I don't see how they would apply when working with monthly sums. So I think I'd suggest removing these tests again. Sorry.

File: source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py
Comment (BjarkeMeier) [reply source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:1]: Looking at the aggregated columns that are not currently tested:
```python
            first(Colname.charge_tax).alias(Colname.charge_tax),
            first(Colname.unit).alias(Colname.unit),
            min(Colname.observation_time).alias(Colname.charge_time),
```
I don't understand why it is valid to pick what just happens to be the first charge_tax and unit? What if there are different values for each in the month?

Is min observation time correct? What if there are no tariff prices for the first day of the month?

Comment (BjarkeMeier) [inline source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:1]: Regarding these two select from the sut:
```python
            lit(None).alias(Colname.metering_point_type),
            lit(None).alias(Colname.settlement_method),
```
I believe they must be null and thus it would be worth adding tests that they are in fact null.

Comment (johevemi) [reply source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:1]: charge_tax is always the same for all tariff in the month,
unit is always the same for tariff (kWh)
min(observerationtime) might be an issue i will look in to that

Comment (johevemi) [reply source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:1]: yes metering_point_type and settlement_method is always null for monthly

Comment (BjarkeMeier) [reply source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:1]: That information would be nice as code doc in the production code 😄

Comment (johevemi) [reply source/databricks/calculation-engine/tests/calculation/wholesale/test_tariff_calculators.py:1]: fixed the issue with min(obstime) and added comments to tax and unit

File: source/databricks/calculation-engine/package/calculation/wholesale/wholesale_calculation.py [hunk 9]
```diff
@@ -98,18 +85,27 @@ def _get_production_and_consumption_metering_points(
...
+            F.from_utc_timestamp(F.col(Colname.observation_time), time_zone)
+        ),
+    )
+    df = df.withColumn("year", F.year(df[Colname.local_date]))
+    df = df.withColumn("month", F.month(df[Colname.local_date]))
+    df = df.withColumn(
+        "first_of_month",
+        F.to_utc_timestamp(
+            F.to_timestamp(F.date_trunc("month", Colname.local_date)), tz=time_zone
+        ),
```
Comment (BjarkeMeier) [inline source/databricks/calculation-engine/package/calculation/wholesale/wholesale_calculation.py:106]: Wouldn't it be simpler and faster (runtime) to simply retrieve it from the arguments passed to the pyspark job?

**Review by BjarkeMeier — APPROVED** [review]
